<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Locating the causal mutation, whilst testing my algorithm on my model genome</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Locating the causal mutation, whilst testing my algorithm on my model genome</h1>

<p>Script: <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/find_causal_mutation.rb">find_causal_mutation.rb</a></p>

<p>Running <code>ruby find_causal_mutation.rb dataset run last_generation</code> will print out the position of the homozygous SNP, at the centre of the peak in homozygous to heterozygous SNP ratio, for both the ordered genome and the fittest permutation outputted from the genetic algorithm. <code>dataset</code> is the directory within <code>fragmented_genome_with_snps/arabidopsis_datasets</code> that contains the input data (FASTA and VCF files) for the algorithm. <code>run</code> is the directory within <code>dataset</code> that contains the output from one run of the algorithm. <code>last_generation</code> should be an integer of the final generation in the run.</p>

<h1>Locating the causal mutation from real world data (and how the location works)</h1>

<p>After running my genetic algorithm with real world data, the causal mutation could be found using a modified version of <code>find_causal_mutation.rb</code>. Use the methods <code>find_peak</code> and <code>closest_snp</code> drom the <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/locate_mutation.rb">LocateMutation</a> class as shown below:</p>

<pre><code class="ruby">peak =  LocateMutation.find_peak(hyp, n) # Find the peak in the approximated (hypothetical SNP) distribution
causal_mutation = LocateMutation.closest_snp(peak, hm)
</code></pre>

<p><code>hyp</code> is a hypothetical distribution that represents the ratio of homozygous to heterozygous SNPs (see <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/snp_dist.rb">SNPDist.hyp_snps</a>) across the genome/permutation.
<code>n</code> is the number of equally spaced points at which the density of <code>hyp</code> SNPs are estimated, and should be a power of two.
<code>hm</code> is the distribution of homozgous SNPs across the genome/permutation, that was used with the heterozygous distribution to create <code>hyp</code>.</p>

<p>The <code>causal_mutation</code> therefore, is the homozgous SNP that lies closest to the highest peak in the ratio of homozgous to heterozgous SNP density.</p>

</body>

</html>

