<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Genetic algorithm flowchart</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Genetic algorithm flowchart</h1>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/evolve.png?raw=true" alt="Image"/></p>

<h3>FASTA file to population</h3>

<p><strong>(1-2)</strong> <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/reform_ratio.rb">ReformRatio::fasta_array</a> obtains A FASTA object for each contig, and adds them to an array. <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/GATOC.rb">GATOC::initial_population</a> randomly orders the FASTA object array, and adds it to a super-array (the population). This is repeated to get the population of the desired size.</p>

<h3>VCF file to SNP positions on each contig</h3>

<p><strong>(3-4)</strong> <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/reform_ratio.rb">ReformRatio::get_snps</a> creates arrays of the SNP positions on each fragment, and outputs arrays and hashes of the other information in the VCF. <em>All this information is stored in memory.</em></p>

<h3>SNP positions for a contig permutation</h3>

<p><strong>(5-6)</strong> <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/reform_ratio.rb">ReformRatio</a> contains other methods that obtain arrays of the homozygous and heterozygous SNPs across the entire genome, for each contig permutation. <em>Perhaps these methods can be grouped into one method, to avoid large arrays of the data being stored in memory</em></p>

<h3>Calculating the ratio of homozygous to heterozygous SNPs, for each permutation</h3>

<p><strong>(7-8)</strong> <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/fitness_score.rb">FitnessScore::count</a> is used to get the number of SNPs at a set number of divisions across the genome. This is done for homozygous and heterozygous SNPs. Then <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/fitness_score.rb">FitnessScore::ratio</a> creates an array of the ratio of homozygous to heterozygous SNPs for each of the genome divisions.</p>

<h3>Permutation fitness</h3>

<p><strong>(9-10)</strong> <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/lib/fitness_score.rb">FitnessScore::score</a> calculates the &ldquo;fitness&rdquo; of a permutation, based on the comparison of it&#39;s distribution of homozygous to heterozygous SNP ratio, with an expected distribution. Currently (June 25th) I am using the distribution from correctly ordered contigs as the expected distribution, since I am still testing whether the genetic algorithm aproach will work.
THEN SAY HOW METHOD WORKS</p>

</body>

</html>

