<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Results: Genetic algorithm run with SNP distances fitness method</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Results: Genetic algorithm run with SNP distances fitness method</h1>

<p>Each facet/panel in the plots below contain a group of replicates with the same parameters:</p>

<ul>
<li>Population size: number of permutations in each population (a new population for every generation)</li>
<li>Select number: number of permutations from each population to create a new population from</li>
<li>Chunk mutants: number of chunk mutants (<a href="https://github.com/edwardchalstrey1/pmeth">PMeth gem</a>) created from randomly chosen permutations, from the fittest selection of the previous generation</li>
<li>Swap mutants: number of swap mutants (<a href="https://github.com/edwardchalstrey1/pmeth">PMeth gem</a>) created from the best permutation from the fittest selection of the previous generation</li>
<li>Save: number of the fittest permutations to carry over from each generation to the next unaltered</li>
<li>Random: number randomly ordered permutations in each generation</li>
</ul>

<style>
table,th,td
{
border:1px solid black;
}
</style>

<table>

  <tr><th>Param</th><th>Population size</th><th>Select number</th><th>Chunk mutants</th><th>Swap mutants</th><th>Save</th><th>Random</th></tr>
  
  <tr> <td>p1</td> <td>100</td> <td>50</td> <td>35</td> <td>35</td> <td>25</td> <td>5</td> </tr>
  <tr> <td>p2</td> <td>20</td> <td>10</td> <td>7</td> <td>7</td> <td>5</td> <td>1</td> </tr>
  <tr> <td>p3</td> <td>50</td> <td>25</td> <td>20</td> <td>20</td> <td>6</td> <td>4</td> </tr>
  <tr> <td>p4</td> <td>20</td> <td>4</td> <td>9</td> <td>9</td> <td>1</td> <td>1</td> </tr>

</table>

<h2>Fitness</h2>

<p>The figure below shows the results of 40 replicates of the genetic algorithm with the parameter groupings in the table above. Although there appears to be some initial improvement in terms of fitness score, all the replicates level out within 60 generations.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/max_density/arabidopsis_datasets/10K_dataset4a/umbrella_plot_fits_total_snp_distances.png?raw=true" alt="Image"/></p>

<p>The figure below shows the same results as the one above, but with the addition of the correct permutation&#39;s fitness. This is shown on the line created from &ldquo;replicate C&rdquo;. Clearly none of the replicates have produced permutations with fitness scores close to that of the correctly ordered genome.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/max_density/arabidopsis_datasets/10K_dataset4a/umbrella_plot_fits_total_snp_distance_with_correct.png?raw=true" alt="Image"/></p>

<h2>Permutation Distance Metrics</h2>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22dev%22%5D.png?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22square%22%5D.png?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22ham%22%5D.png?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22r_dist%22%5D.png?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22lcs%22%5D.png?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/umbrella_plot_%5B%22kt%22%5D.png?raw=true" alt="Image"/></p>

<h2>Looking closer at the changes in SNP distribution for p_run39 replicate</h2>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/p_run39/images_hyp.gif?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/p_run39/images_hm.gif?raw=true" alt="Image"/></p>

<p>The figure below is a representation of how the contig order changes for the best (fittest) permutations in generations of the algorithm.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/10K_dataset4a/circos_p_run39.png?raw=true" alt="Image"/></p>

<h3>Location of causal mutation</h3>

<p>Positions of the homozygous SNP, at the centre of the peak in homozygous to heterozygous SNP ratio:</p>

<ul>
<li>Location of causal mutation in correctly ordered genome: 10156487</li>
<li>Candidate SNP position in best permutation from p_run39: 6105991</li>
</ul>

<h2>Small Dataset</h2>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/small_dataset2final/snp_distance/images_hyp.gif?raw=true" alt="Image"/>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/small_dataset2final/snp_distance/images_hm.gif?raw=true" alt="Image"/></p>

<h3>Location of causal mutation</h3>

<p>Positions of the homozygous SNP, at the centre of the peak in homozygous to heterozygous SNP ratio:</p>

<ul>
<li>Location of causal mutation in correctly ordered 2Kb sequence: 1016</li>
<li>Candidate SNP position in best permutation from this run:      928</li>
</ul>

</body>

</html>

