<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Rearrangement method ideas</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Rearrangement method ideas</h1>

<h2>Machine learning</h2>

<p>Learning involves giving the machine examples, then the more examples, the better the machine gets at telling how a new thing should be catagorized. For example, a machine learning system could be trained on email messages to learn to distinguish between spam and non-spam messages.</p>

<h2>Hill Climbing algorithms</h2>

<p>An arbitrary solution to a problem is generated, then an element of the solution is changed, if this produces a better solution, it is kept. Iteratively, the solution gets better. It can however, reach a local maxima, where changing one element cannot make a better solution, whilst the global maxima (the best possible solution) is not actually reached.</p>

<h2>What to do</h2>

<p>Looking at the <a href="http://bar.utoronto.ca/ngm/cgi-bin/emap.cgi">NGM website</a>, with sample data, we can see that there is a noticeable homo/hetero signal distribution around the causative mutation.</p>

<p>You need to modify your model genome:</p>

<ul>
<li>Two sets of SNPs (hetero and homozygous), that will generate a normally distributed (or similar curve to NGM sample) ratio distribution</li>
<li>The ratio is maximized at the causative mutation, and distributed around it</li>
<li>The standard deviation/ curve of the distribution is down to the kernel density. How tightly does this fit around the data? You decide: we want the distribution to look as normal as possible, as this will be what we try and regenerate when putting the fragments back together</li>
<li>Use a qq plot to make sure the ratio is a normal distribution</li>
<li>then use the qq plot to test the distribution of rearranged fragments&#39; hom/het ratio (score the rearrangements based on how close to the norm dist they are</li>
</ul>

<h1>Questions</h1>

<p>Back-crossing experiment produces F2 EMS mutants, with SNPs distibuted around causative mutation.
Sequencing and contig assembly produces fragments with SNPs at known positions.</p>

<ol>
<li><p>What will the exact function of my tool be?</p>

<ul>
<li>Locate the causative mutation by looking at the fragments</li>
<li>The SNPs will form a homozygous/heterozygous ratio that is normally distributed about the causative mutation?</li>
</ul></li>
<li><p>How should I model this? </p>

<ul>
<li>If I put &ldquo;homozygous&rdquo; and &ldquo;heterozygous&rdquo; SNPs into the fasta/vcf data, how will I then re-identify them? NGM uses chastity statistic to measure the frequencies of hom and het SNPs from pooled data. But the experiment I&#39;m modelling is different? If not, then will I not just be replicating NGM?</li>
<li>I have come up with a way of making a hom/het ratio distribution, <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/ratio/ratio.R">see ratio.R</a></li>
<li>Once I have made new datasets for this version of the model genome, I can read off generated fasta/vcf files and put frags in correct order and write some code that can work out/display the ratio distribution. This code will be used later when rearranging randomly ordered fragments.</li>
</ul></li>
<li><p>How will the algorithm work?</p>

<ul>
<li>I want to read off the data e.g. fasta/vcf and regenerate the normal distributed ratio, which will be done by putting the fragments back into the correct order</li>
<li>The fragments will be rearranged by a hill climbing algorithm: this will iteratively attempt to reorder the fragments in a way that creates a hom/het SNP ratio normal distribution (fragments with zero ratio need not be ordered correctly/ may be ignored completely)</li>
<li>The causative mutation will be at the centre of the distribution: see figure source <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/ratio/ratio.R">see ratio.R</a>. The distribution is reformed by reading the SNP positions from the fragments, and working out their SNP position in the genome (for this fragment order). The SNP lists, hom and het, can then be used to work out the ratio distribution.</li>
</ul></li>
</ol>

<h2>Hill Climbing</h2>

<h3>A reasonable starting solution is required:</h3>

<p>For my fragment rearrangement algorithm I propose using my skew/gradient method. Even if this method proves to be of equal uselessness for reforming the narrow ratio distribution, than the wider SNP distribution I have tested it on so far, It can at least split the fragments into 2 groups, and then they can be ordered by SNP density. One problem I may come across is that <strong>ordering the fragments by SNPs per nucleotide (density) won&#39;t be very useful for low SNP numbers (within the &ldquo;non-recombinant region&rdquo; where there are just homozygous SNPs)</strong>. However, the low number of SNPs does mean that the normal distribution will almost definitely only be reformed by a correct rearrangement.</p>

<h3>Evaluate correctness funtion</h3>

<p>This would be an extension of the code I came up with earlier that calculates the ratio distribution (for the fragment rearrangement), but would score it on how normal it is. I could use a qq plot to do this.</p>

<h3>Operator to modify complete solution</h3>

<p><em>Thinking about it, for this to work, we will want to deal with all of the fragments, some of them from outside the recombinant region may have a noticeable low sub-peak of ratio (i.e. bumps in the overall ratio plot). To make the model genome more realistic, you may want to add some more of these &ldquo;bumps&rdquo;.</em></p>

<ul>
<li>Hopefully the initial solution will reveal some distinct peaks</li>
</ul>

</body>

</html>

