<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Rearrangement methods part 2</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Rearrangement methods part 2</h1>

<p>Having created a new dataset (<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/dataset2.md">dataset2</a>), I will re-run the existing fragment rearrangement methods described in <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/rearrangement_methods.md">rearrangement_methods.md</a>. I will then continue to write up additional experiments here. I used this <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/fasta_vcf_d2/frags_shuffled.fasta">fasta</a> and this <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/fasta_vcf_d2/snps.vcf">VCF</a>. Whilst dataset 1 contained 1310 fragments, dataset 2 contains 1321. </p>

<h2>Existing Method Scores</h2>

<p>Below are a list of the methods used in <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/rearrangement_methods.md">rearrangement_methods.md</a>, and a table comparing the scores for dataset 1 and 2:</p>

<p>Highest possible score (C0), Density order Score (C1), Random Score (C2), Even Odd Method Score (M1a), Odd Even Method Score (M1b), Left Right Method Score (M2a), Left Right Density Method Score (M2b)</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/dataset_scores_table.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/dataset_scores_table.png">Figure 1</a></p>

<p>As you can see from Figure 1, for each rearrangement method, the ordinal similarity score is similar in both datasets (and slightly higher for dataset 2 in most cases, probably due to the larger number of fragments). This was what I expected for the controls and method 1 (odd/even), since the change in the model genome was aimed at maximising the potential of method 2b (left right density), see <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/dataset2.md">dataset2</a>. This method relies on the skew of SNPs to determine each fragments position in the rearranged order, see <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/rearrangement_methods.md">rearrangement_methods.md</a>.</p>

<p>Figure 1 also shows that method 2b has a similar ordinal similarity score for both datasets. What this may suggest is that the skew of SNPs on a fragment is in fact not a good indicator of its position in the genome. However, it could be that the current skew method (2b) is simply incorrect in its placing of the fragments.</p>

<p>For SNP density plots from dataset 2 (and bar plot of scores) see <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/tree/master/figures/figures_d2">dataset 2 plots.</a></p>

<p>To investigate how each fragments&#39; skew relates to it&#39;s position in the proper order, I have plotted graphs of the &ldquo;SNP gradient&rdquo; for each fragment. The R script used to create figure 2, 3 and 4 can be found <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/skew_scatta.R">here</a>. 
My prediction was that the gradients of frags would show a similar pattern to the normal distribution of SNPs, with the gradients being higher towards the midpoint of the model genome (where the mutation is), and lower at either end. However, I also thought that some of the fragments right in the centre of the genome might have a lower gradient, due to their having many SNPs across their lengths (i.e. on either side, reducing the skew). In figure 2 below, I use an absolute value of gradient. It will also be useful to show the difference between the positive and negative gradient values,  to determine whether the gradient (and skew) can be used to tell the side of the model genome (either side of the mutant point) that a fragment is from (figure 4).</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/skew_scatter.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/skew_scatter.png">Figure 2</a>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/skew_scatter2.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/skew_scatter2.png">Figure 3</a>
<img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/example_gradient_f629.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/example_gradient_f629.png">Figure 4</a></p>

<p>Figures 2 and 3 are plots of each fragment in their proper order, against their gradients. In figure 2 the gradients used are the absolute values of the gradients used in figure 3. </p>

<p>The &ldquo;SNP gradients&rdquo; are determined by creating plots of the nucleotides against the number of SNPs for each fragment, where the number of SNPs is a value of 0 or 1.
Figure 4 shows an example plot, where the gradient of the blue line (a linear model fitted to the data), is the gradient for that fragment plotted in figure 2/3. For each fragment, the skew of SNPs to either side of the fragment produces a positive gradient (right skew) or negative gradient (left skew, as with fragment629 in figure 2). Calculating the skew in a way that could be quantified was important, in order to plot figure 2/3.</p>

<p>Interpreting what figure 2 and 3 show about the SNP gradient/skew of fragments is difficult, because of the large number of fragments that have a gradient of zero (or close to zero). As you can see, the loess smoothing lines (red) for figure 2 and 3 are drawn across value zero of gradient. The reason for this is that many of the fragments have either 0 SNPs (which gives them a gradient of exactly zero) or 1 SNP, or another low number of SNPs. </p>

<p>This is another problem with my model genome that I have identified; the gradients and skew may reveal more about the position of fragments if a larger number of SNPs is used. Fragments with just 1 SNP (or any low number) are less likely to have a useful skew/gradient than fragments with a larger number of SNPs. If the SNPs in my model genome are not very concentrated with regard to the fragment size, the position of SNPs on individual fragments could be a lot more arbitrary. If the fragments were larger (in proportion to the genome size) and/or the genome had a higher number of SNPs around the mutant position, the fragments may exhibit a much clearer skew, that is related to their position in the genome. Figure 5 below illustrates this point.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/example_gradient_f321.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/figures/figures_d2/example_gradient_f321.png">Figure 5</a></p>

<p>Working out the gradient in the way I have done may not be useful for fragment rearrangement. However, it does show that the skew method (method 2b) is not likely to be effective with the model genome used to create datasets 1 and 2.</p>

<p>Dataset 1 and 2 are from a model genome of 200Kb in length with 1000 SNPs normally distributed around the mutant position 100,000, and fragment sizes of 50-250b.</p>

<p>At this stage, I have decided to create a more realistic model genome based on <em>Arabidopsis thaliana</em> chromosome 4, see <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/dataset3_arabidopsis.md">here</a> for details.</p>

</body>

</html>

