<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Rearrangement methods part 4</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>





</head>

<body>
<h1>Rearrangement methods part 4</h1>

<h2>Testing the SNP Skew</h2>

<p>Since the Kernel density estimate was not normalized for each fragment, I have come up with a new way of testing SNP skew as a method for rearranging fragments with SNPs into the correct order. See the project <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/README.md">README</a> and parts <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/rearrangement_methods.md">1</a>, <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/p2_rearrangement_methods.md">2</a> and <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/p3_rearrangement_methods.md">3</a> of rearrangement methods for details.</p>

<p>The new method I have used for determining SNP gradients for each of the fragments, relies on a different way of representing SNP density across the fragment, which I will refer to as the distance method. Figures 1 and 2 show the gradients worked out by this method, plotted for each of the fragments in one of my <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/writeup/arabidopsis_chromosome4.md"><em>Arabidopsis</em> datasets</a> (dataset5). The ruby script used to generate all the figures is <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/skew_scatta.rb">here</a>; it uses an <a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/skew_scatter.R">R script</a> and the RinRuby gem to make the plots.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_1_10000_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_1_10000_d1_m0.5.png">Figure 1</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_1_10000_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_1_10000_d1_m0.5.png">Figure 2</a></p>

<p>Figures 1 and 2 show that the gradients are slightly higher for fragments from the middle of the model genome, where the centre of the normal distribution of SNPs is.</p>

<p>The distance method works by measuring the distance of each base from the nearest SNP in the contig. A constant (in this case 10,000) is divided by the the distance value so that the &#39;density&#39; decreases with the inverse of an increasing arithmetic progression, the further a base is from the closest SNP. The &#39;density&#39; values attributed to the SNP&#39;s themselves are 2x the constant (20,000). Figures 3 and 4 are examples of gradient determination using the distance method, the gradient lines are shown in blue.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f681_mins1_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f681_mins1_d1_m0.5.png">Figure 3</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f258_mins1_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f258_mins1_d1_m0.5.png">Figure 4</a></p>

<p>Since the causative mutation is likely to be on a fragment with a high SNP density, I have repeated the plots of figure 1 and 2 for just the fragments with 30+ SNPs (figures 5 and 6), to see if the same pattern is shown. The pattern is the same (there basically isn&#39;t one), all that figures 5 and 6 show is that the fragments with higher SNP numbers tend to be at the centre of the distribution, but I already knew that (e.g. fragments 350-950).</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_30_10000_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_30_10000_d1_m0.5.png">Figure 5</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_30_10000_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_30_10000_d1_m0.5.png">Figure 6</a></p>

<p>Figure 7 shows that using the skew for fragment rearrangement and causative mutation detection may not be useful, since <strong>some</strong> of the fragments with the highest SNP densities (where the mutation is most likely to be found) have very low gradients, therefore their SNP skew is not very likely to be a useful indicator of the region of the model genome a fragment comes from.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f729_mins30_d1_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f729_mins30_d1_m0.5.png">Figure 7</a></p>

<h3>Distance method 2</h3>

<p>Since the results of the first distance method showed that the skew is unlikely to be useful, I took a different approach to be sure:</p>

<p>The second distance method works by measuring the distance of each base from the nearest SNP in the contig. The distance value is taken from a constant (in this case 20,000) so that the &#39;density&#39; decreases with an arithmetic progression, the further a base is from the closest SNP. The &#39;density&#39; values attributed to the SNP&#39;s themselves are the constant. Figures 8 and 9 are examples of gradient determination using the distance method 2, the gradient lines are shown in blue.</p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_1_20000_d2.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_1_20000_d2.png">Figure 8</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_1_20000_d2.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_1_20000_d2.png">Figure 9</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f258_mins1_d2_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f258_mins1_d2_m0.5.png">Figure 10</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f681_mins1_d2_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f681_mins1_d2_m0.5.png">Figure 11</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_30_20000_d2.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_abs_30_20000_d2.png">Figure 12</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_30_20000_d2.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/skew_scatter_grad_30_20000_d2.png">Figure 13</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f729_mins30_d2_m0.5.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/example_gradient_f729_mins30_d2_m0.5.png">Figure 14</a></p>

<h3>Visualizing the SNPs on fragments</h3>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f258.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f258.png">Figure 15</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f681.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f681.png">Figure 16</a></p>

<p><img src="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f729.png?raw=true" alt="Image"/>
<a href="https://github.com/edwardchalstrey1/fragmented_genome_with_snps/blob/master/arabidopsis_datasets/dataset5/figures/how_scatta_f729.png">Figure 17</a></p>

</body>

</html>

